File: ImpulseGenerator.java
import processing.core.PApplet;

public class ImpulseGenerator extends Generator {

	int punchPointMillis;
	int duty;
	int echo;
	float echoDecay;

	public ImpulseGenerator(PApplet p, String name, float amp, int period, int duty, int echo, float echoDecay) {
		super(p, name, amp, 0, period);
		this.duty = duty;
		this.echo = echo;
		this.echoDecay = echoDecay;
	}

	public void punch(int millis) {
		punchPointMillis = millis;
	}

	public void setEcho(int echo, float echoDecay) {
		this.echo = echo;
		this.echoDecay = echoDecay;
	}

	public void setDuty(int duty) {
		this.duty = duty;
	}

	public float value(int millis) {
		int ellapse = millis - punchPointMillis;
		int count = ellapse / period;
		if (count > echo) return 0.0f;

		if (ellapse % period > 0 && ellapse % period < duty) {
			return amp * p.pow(echoDecay, count);
		} else {
			return 0.0f;
		}
	}

}

File: python/parquette-lights/src/parquette/lights/generators/impulse_generator.py
from .generator import Generator
from ..category import Category
from ..osc import OSCManager


class ImpulseGenerator(Generator):
    STANDARD_ATTRS = ["amp", "duty"]

    punch_point: float = 0
    _punch_pending: bool

    def __init__(
        self,
        *,
        name: str,
        category: Category,
        amp: float = 1,
        offset: float = 0,
        duty: float = 100,
    ):
        super().__init__(
            name=name, category=category, amp=amp, offset=offset, period=1, phase=0
        )

        self.duty = duty
        self._punch_pending = False

    def punch(self) -> None:
        # Defer the actual punch start to the next value() call so that the
        # impulse begins exactly at the start of the next runChannelMix tick.
        # Otherwise a punch fired between ticks can land partway through (or
        # entirely past) a short duty window and be missed by the mix loop.
        self._punch_pending = True

    def register_punch(self, osc: OSCManager) -> None:
        """Register /gen/ImpulseGenerator/{name}/punch on the dispatcher."""
        addr = "/gen/{}/{}/punch".format(type(self).__name__, self.name)
        osc.dispatcher.map(addr, lambda _addr, *args: self.punch())

    def value(self, millis: float) -> float:
        if self._punch_pending:
            self.punch_point = millis
            self._punch_pending = False

        ellapsed: float = millis - self.punch_point

        if ellapsed < self.duty:
            return self.amp + self.offset
        else:
            return self.offset
